motorApp/SM300Src/SM300Driver.cpp

```diff
@@ -0,0 +1,583 @@
+}
+
+/** Send a querry string to the controller and get a return value.
+  * Querry string is prefeced with ACK STX and postfixed with EOT
+  * return string ends with ETX
+  * \param[in] querry the querry to send
+  * \param[in] what form the querry reply has; true if the reply has EOT false for ETX with BCC
+  * \returns success of write and read of querry string
+  */
+asynStatus SM300Controller::sendQuerry(const char * querry, bool hasEotEnding) {
+	if (hasEotEnding) {
+		setTerminationChars("\x04", 1, "\x04", 1);
+	} 
+	else {
+		setTerminationChars("\x03", 1, "\x04", 1);
+	}
+	//send data format 2
+	sprintf(this->outString_, "\x06\x02%s", querry);
+	return this->writeReadController();
+}
+
+/** Send a command string to the controller.
+* Command string is prefeced with ACK STX and postfixed with EOT
+* \param[in] command the command to send
+* \returns success of write and read of acknowledgement from controller
+*/
+asynStatus SM300Controller::sendCommand(const char * command) {
+	setTerminationChars("\x06", 1, "\x04", 1);
+	//send data format 2
+	sprintf(this->outString_, "\x06\x02%s", command);
+	return this->writeReadController();
+}
+
+/** 
+  * Return true if the controller registering moving motors
+  */
+bool SM300Controller::is_moving() {
+	return is_moving_;
+}
+
+/**
+  * deal with db records being set which are integers
+  * \returns status
+  */
+asynStatus SM300Controller::writeInt32(asynUser *pasynUser, epicsInt32 value) {
+	asynStatus status;
+
+	int function = pasynUser->reason;		//Function requested
+	if (function == reset_) {
+		if (value == 0) return asynSuccess;
+		status = perform_reset();
+		callParamCallbacks();
+	}
+	else if (function == disconnect_) {
+		setIntegerParam(disconnect_, 1);
+		setTerminationChars("\x06", 1, "\x04", 1);
+		sprintf(this->outString_, "\x06\x02%s", "M77");
+		status = writeController();
+		setIntegerParam(disconnect_, 0);
+		callParamCallbacks();
+	}
+	else if (function == reset_and_home_) {
+		setIntegerParam(reset_and_home_, 1);
+		callParamCallbacks();
+		perform_reset();
+		for (int i = 0; i < numAxes_; i++) {
+			getAxis(i)->home(0, 0, 0, 0);
+		}
+		setIntegerParam(reset_and_home_, 0);
+		callParamCallbacks();
+	}
+	else {
+		status = asynMotorController::writeInt32(pasynUser, value);
+	}
+	return status;
+}
+
+/**
+ * Perform a reset of the controller sending it initiation parameters
+ * \returns status
+*/
+asynStatus SM300Controller::perform_reset() {
+	setIntegerParam(reset_, 1);
+	callParamCallbacks();
+	setTerminationChars("\x04", 1, "\x04", 1);
+
+	//send empty string with ACK to clear the buffer		
+	asynStatus status = this->writeController();
+
+	//set termination character is EOT without check sum *
+	// when sending send <CR> incase this mode is switched on
+	setTerminationChars("\x06", 1, "\x04\x0D", 2);
+	if (sendCommand("PEK0")) return status;
+
+	setTerminationChars("\x06", 1, "\x04", 1);
+	// set achknowlegement on
+	if (sendCommand("PEL1")) return status;
```

Comment:
Is there a reason for not doing this as a loop? Much easier if you ever want to change what to do for each command.

```
const char * commands[] = {
    "PEL1",
    "B/ G01",
    ...
};
int n_array = (sizeof (commands) / sizeof (const char *));
for (i = 0; i < n_array; i++) {
    if (sendCommand(commands[i]) return status;
}
```